src/functions/tester/types.ts

```diff
@@ -0,0 +1,64 @@
+import type {
+  ParameterSetDefinition,
+  PossibleParameterKeys,
+} from "../../parameters/mod.ts";
+import type { SlackFunction } from "../mod.ts";
+import type { FunctionContext, FunctionRuntimeParameters } from "../types.ts";
+
+export type SlackFunctionTesterArgs<
+  InputParameters,
+> =
+  & Partial<
+    FunctionContext<InputParameters>
+  >
+  & {
+    inputs: InputParameters;
+  };
+
+export type CreateContext<
+  InputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+> = {
+  (
+    args: SlackFunctionTesterArgs<
+      FunctionRuntimeParameters<InputParameters, RequiredInput> | undefined
+    >,
+  ): FunctionContext<
+    FunctionRuntimeParameters<InputParameters, RequiredInput>
+  >;
+};
+
+export type SlackFunctionTesterResponse<
+  InputParameters extends ParameterSetDefinition,
+  RequiredInput extends PossibleParameterKeys<InputParameters>,
+> = {
+  createContext: CreateContext<InputParameters, RequiredInput>;
+};
+
+// Slack Function Tester is overloaded to accept either a string or a SlackFunction
+export type SlackFunctionTesterFn = {
+  // Accept a Slack Function
+  <
+    InputParameters extends ParameterSetDefinition,
+    OutputParameters extends ParameterSetDefinition,
+    RequiredInput extends PossibleParameterKeys<InputParameters>,
+    RequiredOutput extends PossibleParameterKeys<OutputParameters>,
+  >(
+    funcOrCallbackId: SlackFunction<
+      InputParameters,
+      OutputParameters,
+      RequiredInput,
+      RequiredOutput
+    >,
+  ): SlackFunctionTesterResponse<
+    InputParameters,
+    RequiredInput
+  >;
+
+  // Accept a string
+  (funcOrCallbackId: string): {
+    createContext: {
+      <I>(args: SlackFunctionTesterArgs<I>): FunctionContext<I>;
+    };
+  };
```

Comment:
This is the trickiest bit of this PR. Using `SlackFunctionTesterResponse` like above I would run into a problem when I tried to use a handler typed to our `BaseSlackFunctionHandler` and pass the result of `createContext` into that handler.

This was the best way I found to infer the type passed to a `createContext` call that worked for both `BaseSlackFunctionHandler` and `SlackFunctionHandler`. Open to any ideas on how this could be improved

Reply:
Umm, i wouldn't have even come up with this idea, so def. don't have other ideas. It works amazingly though!